### ISSUE TITLE
libtorch for Windows. MNIST example does no work.

### ISSUE THREAD

**Issue by alex-zador** [pytorch/pytorch#35713]
I've tried to run mnist train example (https://github.com/pytorch/examples/tree/master/cpp/mnist)
Code crushes on train_dataset creation whith message "Unhandled exception at 0x00007FFE7B879179 in MNIST.exe: Microsoft C++ exception: c10::Error at memory location 0x00000068D2EFF4B0. occurred".
[auto train_dataset = torch::data::datasets::MNIST(kDataRoot)
        .map(torch::data::transforms::Normalize<>(0.1307, 0.3081))
        .map(torch::data::transforms::Stack<>());]
Environment:
Win10, VS 2019, Release, CPU.

I'm not sure if mnist-data prpared correctly. Just put them to ./data catalog
26.03.2020  12:10         1 648 877 t10k-images-idx3-ubyte.gz
26.01.1998  18:07         7 840 016 t10k-images.idx3-ubyte
26.03.2020  12:10             4 542 t10k-labels-idx1-ubyte.gz
26.01.1998  18:07            10 008 t10k-labels.idx1-ubyte
26.03.2020  12:09         9 912 422 train-images-idx3-ubyte.gz
18.11.1996  18:36        47 040 016 train-images.idx3-ubyte
26.03.2020  12:09            28 881 train-labels-idx1-ubyte.gz
18.11.1996  18:36            60 008 train-labels.idx1-ubyte

What is wrong?

**Comment by ailzhang** [comment]
Not sure who to cc for the tutorial, cc: @jlin27 ?

**Comment by peterjc123** [comment]
Please check the following items:
1. The configuration of the project and the libtorch library is the same.
2. The data is put in the right directory and they are not corrupted.
3. The DLLs are copied to the directory of your executable.

**Comment by peterjc123** [comment]
Actually we could use this property for the out-of-box experience for VS.
https://stackoverflow.com/questions/23950887/does-cmake-offer-a-method-to-set-the-working-directory-for-a-given-build-system

**Comment by alex-zador** [comment]
@peterjc123, I have 4 projects in a single VS solution, all configared the same way (copy dlls, include and libs, x64, Release). 3 projects work fine. So I think every thing is OK with configuration and dlls.

Data (gz & unzipped) are in ./data folder in project dir. Project dir is set as Working Directory. I can open a stream fistream  ifs("./data/t10k-labels.idx1-ubyte") and ifs.is_open() returns true.
One of my project converts mnist data to set of PNG and info.txt. So I think them are not corrupted.

I use prebuilt libraries for Windows...  

So, I have no ideas on the reason of crush...

**Comment by peterjc123** [comment]
@alex-zador Wha about compiling it against Debug? You may get more info about it using the PDBs.

**Comment by alex-zador** [comment]
Nothing new... 
Unhandled exception at 0x00007FFE7B879179 in MNIST.exe: Microsoft C++ exception: c10::Error at memory location 0x000000EA232FE4F0. occurred

Could project archive help?

**Comment by peterjc123** [comment]
@alex-zador Have you copied the PDBs to the location of the executable?

**Comment by alex-zador** [comment]
@peterjc123. Shure! Postbuild event
xcopy C:\libtorch\libd\*.dll  "$(TargetDir)"  /d
xcopy C:\libtorch\libd\*.pdb  "$(TargetDir)"  /d

I also checked that visually...
Directory of C:\Users\alex\Desktop\...\x64\debug

Directory of C:\Users\alex\Desktop\...\x64\debug

31.03.2020  09:39    <DIR>          .
31.03.2020  09:39    <DIR>          ..
31.03.2020  08:58         1 047 040 c10.dll
31.03.2020  08:58         7 606 272 c10.pdb
31.03.2020  08:58           348 160 caffe2_module_test_dynamic.dll
31.03.2020  08:58         6 721 536 caffe2_module_test_dynamic.pdb
31.03.2020  08:58         1 717 584 libiomp5md.dll
31.03.2020  08:58           111 952 libiompstubs5md.dll
31.03.2020  09:26           850 432 MNIST.exe
31.03.2020  09:26             7 389 MNIST.exp
31.03.2020  09:26         3 787 360 MNIST.ilk
31.03.2020  08:47            12 460 MNIST.lib
31.03.2020  09:26         4 796 416 MNIST.pdb
31.03.2020  08:59       488 905 216 torch.dll
31.03.2020  08:59     1 316 564 992 torch.pdb

May be no stacktrace because of unhandled exception?

**Comment by alex-zador** [comment]
Here is Dropbox link for VS 2019 project
https://www.dropbox.com/s/wkb57oq7yc5v8v9/MNISTExample.zip?dl=0
libtorch folder described it ReadMe.txt

**Comment by byungheon** [comment]
@alex-zador I have exactly the same issue now. Have you found any solution for this problem?

**Comment by alex-zador** [comment]
> r

No. I used this example for students and just made CustomDataSet from MNIST data. After that everything works fine...

**Comment by nitinshyamk** [comment]
+1 to this issue.

Unfortunately, I'm having the same issue as well with the MNIST dataset, similar setup as described.  I'm using Visual Studio, but with CMake as the build system. If it's just an issue with the MNIST custom data set, I'll go the same route too, thanks for leaving that info @alex-zador  . W/out access to the mnist.cpp file, it's difficult to take a look though I believe it's the constructor or an attempt to call the constructor that's throwing an error.

**Comment by yunseokddi** [comment]
> I figured out this issue @peterjc123. The memory issue is due to a file name misspecification, that's why we were getting a memory issue. Unfortunately, the ifstream creation fails silently and nontransparently.
> 
> Notice that the MNIST files output by unzipping from Yann Lecun's MNIST webpage are of the form
> "t10k-images.idx3-ubyte"
> whereas the code https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/src/data/datasets/mnist.cpp expects something of the form "train-images-idx3-ubyte".
> 
> Although this is certainly simple to fix from a user's perspective my feeling is that the code should be modified to use the expected file names with a ".' , since newcomers to the Pytorch C++ API may find this unexpected.
> Then again, this could break others' code, so curious what admins think. Otherwise, putting documentation somewhere would be helpful.

Thx! I've been looking for this all day.
